### ISSUE TITLE
react/no-unused-class-comp-methods

### ISSUE THREAD

**Issue by sarneeh** [yannickcr/eslint-plugin-react#2166]
I'd like to have a rule to detect unused `React` class methods. The rule would filter out `React` lifecycle methods, take the other ones and check if they are used anywhere in the class itself.

Example:
```js
class Component extends React.Component {
    // error: unused method
    someMethod() {
        // something...
    }

    makeApiCall() {
        // do api call...
    }

    handleClick() {
        this.makeApiCall();
    }

    render() {
         return (<button onClick={this.handleClick}>Do it!</button>);
    }
}
```

The motivation behind this rule is that it's actually a bad practice to reference `React.Component` instance methods outside of the instance itself, so it would make sense to check if any methods inside the class are being used or are some kind of unused leftover.

**Comment by ljharb** [comment]
You can’t possibly know it’s unused statically - anything that imports Component could reference Component.prototype.someMethod or similar.

**Comment by ljharb** [comment]
Similarly, you never directly use, say, componentDidUpdate, but React does - and there’s no way for a general purpose linter to catch things like that.

**Comment by sarneeh** [comment]
> You can’t possibly know it’s unused statically - anything that imports Component could reference Component.prototype.someMethod or similar.

I know that, but AFAIK it's not a good practice to reference `React.Component` static methods outside of the class component itself.

> Similarly, you never directly use, say, componentDidUpdate, but React does - and there’s no way for a general purpose linter to catch things like that.

Exactly - you never directly use those and you probably shouldn't also use other component methods outside of it. Knowing this you can filter out `React`-specific methods (`componentDidUpdate` etc.) and check if other methods are used in the class itself.

**Comment by ljharb** [comment]
I suppose that might make sense for a React class component, sure. maybe `no-unused-class-component-methods`

**Comment by sarneeh** [comment]
@ljharb Yeah! I'm sorry, maybe I wasn't specific enough about the request :)

**Comment by n-filatov** [comment]
Can i take this?

**Comment by ljharb** [comment]
Anything with the help wanted label, without a linked PR, is always up for grabs.

**Comment by pawelnvk** [comment]
I saw that recently there was no action taken regarding this topic. I prepared solution that might help. Let me know, if there is something I should fix there. https://github.com/yannickcr/eslint-plugin-react/pull/2239

**Comment by pawelnvk** [comment]
@ljharb Do you have time for checking, if this PR is helpful?

**Comment by trevorwhealy** [comment]
This would make such a great lint rule! ❤️

**Comment by Kiwka** [comment]
It would so wonderful to have this rule 🙏

**Comment by BigPun86** [comment]
+1

**Comment by jamesroyston** [comment]
and another +1 🙂

**Comment by evegreen** [comment]
+1, thx!

**Comment by mixa9269** [comment]
+1, please

**Comment by Waltari10** [comment]
There already is a separate npm plugin released for this, but unfortunately I didn't get it to work. Here it is, if somebody has better luck with it, let me know: https://www.npmjs.com/package/eslint-plugin-no-unused-react-component-methods

**Comment by orassayag** [comment]
Any update about this issue? It's something that I really need.

**Comment by jakeleventhal** [comment]
bump

**Comment by jakeleventhal** [comment]
when i try to add this rule it appears to not exist

`Definition for rule 'react/no-unused-class-component-methods' was not found  react/no-unused-class-component-method`
`"eslint-plugin-react": "7.26.1"`

**Comment by ljharb** [comment]
@jakeleventhal that's because it's not released yet - if you click to the PR, and to the commit that merged it (https://github.com/yannickcr/eslint-plugin-react/commit/1371f71db8284ac087e115b705bfc6ef5ab6b6fe) you'll see it's not in any tags.
